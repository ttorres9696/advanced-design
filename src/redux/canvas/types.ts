import { Vector2d } from 'konva/types/types';

import { ShapeCoords } from '../../shared/components/ShapeElement/types';

export const SELECT_SHAPE = 'SELECT_SHAPE';
export const DESELECT_SHAPE = 'DESELECT_SHAPE';
export const INCREASE_ZOOM = 'INCREASE_ZOOM';
export const DECREASE_ZOOM = 'DECREASE_ZOOM';
export const NORMALIZE_ZOOM = 'NORMALIZE_ZOOM';
export const SET_CANVAS_ORIGIN = 'SET_CANVAS_ORIGIN';
export const INCREMENT_TOTAL_SHAPES = 'INCREMENT_TOTAL_SHAPES';
export const DECREMENT_TOTAL_SHAPES = 'DECREMENT_TOTAL_SHAPES';
export const SET_NEW_DRAGGABLE_POINT_MODE = 'SET_NEW_DRAGGABLE_POINT_MODE';
export const SET_MOUSE_POSITION = 'SET_MOUSE_POSITION';
export const SET_DELETE_DRAGGABLE_POINT_MODE = 'SET_DELETE_DRAGGABLE_POINT_MODE';
export const TOGGLE_MAGNET_MODE = 'TOGGLE_MAGNET_MODE';
export const SET_DRAGGING = 'SET_DRAGGING';
export const SET_SELECT_MODE = 'SET_SELECT_MODE';

export const originCoords = 20;
export type SelectionType = 'reset' | 'add';

interface SelectShapeAction {
  type: typeof SELECT_SHAPE;
  payload: {
    solarElementId: string;
    layer: string;
    selectionType: SelectionType;
  };
}

interface DeselectShapeAction {
  type: typeof DESELECT_SHAPE;
  payload: {
    solarElementId: string;
  };
}

interface IncreaseZoomAction {
  type: typeof INCREASE_ZOOM;
}

interface DecreaseZoomAction {
  type: typeof DECREASE_ZOOM;
}

interface NormalizeZoomAction {
  type: typeof NORMALIZE_ZOOM;
}

interface SetCanvasOriginAction {
  type: typeof SET_CANVAS_ORIGIN;
  payload: ShapeCoords;
}

interface IncrementTotalShapesAction {
  type: typeof INCREMENT_TOTAL_SHAPES;
}

interface DecrementTotalShapesAction {
  type: typeof DECREMENT_TOTAL_SHAPES;
  payload: number;
}

interface SetNewDraggablePointModeAction {
  type: typeof SET_NEW_DRAGGABLE_POINT_MODE;
  payload: boolean;
}

interface SetDeleteDraggablePointMode {
  type: typeof SET_DELETE_DRAGGABLE_POINT_MODE;
  payload: boolean;
}

interface SetMousePositionAction {
  type: typeof SET_MOUSE_POSITION;
  payload?: Vector2d;
}

interface ToggleMagnetModeAction {
  type: typeof TOGGLE_MAGNET_MODE;
}

interface SetDragging {
  type: typeof SET_DRAGGING;
  payload: boolean;
}

interface SetSelectMode {
  type: typeof SET_SELECT_MODE;
  payload: boolean;
}

export type CanvasActionTypes =
  | SelectShapeAction
  | DeselectShapeAction
  | IncreaseZoomAction
  | DecreaseZoomAction
  | NormalizeZoomAction
  | SetCanvasOriginAction
  | IncrementTotalShapesAction
  | DecrementTotalShapesAction
  | SetNewDraggablePointModeAction
  | SetDeleteDraggablePointMode
  | SetMousePositionAction
  | ToggleMagnetModeAction
  | SetDragging
  | SetSelectMode;
