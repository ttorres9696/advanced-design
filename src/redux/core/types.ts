import { ModuleSpec } from '../../shared/models/SolarDesign.interface';
import { DeleteDialogParams, WindowDimension } from './state';

export const FETCHED_MODULE_SPECS = 'FETCHED_MODULE_SPECS';
export const SET_DELETE_DIALOG_PARAMS = 'SET_DELETE_DIALOG_PARAMS';
export const SET_WINDOW_DIMENSION = 'SET_WINDOW_DIMENSION';
export const SET_PRE_LOADING = 'SET_PRE_LOADING';
export const CANCEL_PRE_LOADING = 'CANCEL_PRE_LOADING';
export const DECREASE_PRODUCTION_VALUE = 'DECREASE_PRODUCTION_VALUE';
export const INCREASE_PRODUCTION_VALUE = 'INCREASE_PRODUCTION_VALUE';
export const RESET_PRODUCTION_VALUE = 'RESET_PRODUCTION_VALUE';

interface FetchedModuleSpecsAction {
  type: typeof FETCHED_MODULE_SPECS;
  payload: ModuleSpec[];
}

interface SetDeleteDialogParamsAction {
  type: typeof SET_DELETE_DIALOG_PARAMS;
  payload: DeleteDialogParams;
}

interface SetWindowDimensionAction {
  type: typeof SET_WINDOW_DIMENSION;
  payload: WindowDimension;
}

interface SetPreLoadingAction {
  type: typeof SET_PRE_LOADING;
  payload: boolean;
}

interface CancelPreLoading {
  type: typeof CANCEL_PRE_LOADING;
}

interface DecreaseProductionValue {
  type: typeof DECREASE_PRODUCTION_VALUE;
  payload: number;
}

interface IncreaseProductionValue {
  type: typeof INCREASE_PRODUCTION_VALUE;
  payload: number;
}

interface ResetProductionValue {
  type: typeof RESET_PRODUCTION_VALUE;
}

export type CoreActionTypes =
  | FetchedModuleSpecsAction
  | SetDeleteDialogParamsAction
  | SetWindowDimensionAction
  | SetPreLoadingAction
  | CancelPreLoading
  | DecreaseProductionValue
  | IncreaseProductionValue
  | ResetProductionValue;
