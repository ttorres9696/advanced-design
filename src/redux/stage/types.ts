import { ShapeCoords } from '../../shared/components/ShapeElement/types';
import { ModuleSpec, SolarDesign } from '../../shared/models/SolarDesign.interface';
import { SolarElement } from '../../shared/models/SolarElement.interface';
import { SelectionType } from '../canvas/types';

export const ADD_SHAPE = 'ADD_SHAPE';
export const UPDATE_SHAPE = 'UPDATE_SHAPE';
export const REMOVE_SHAPE = 'REMOVE_SHAPE';
export const TOGGLE_VISIBILITY = 'TOGGLE_VISIBILITY';
export const TOGGLE_LOCK = 'TOGGLE_LOCK';
export const IMPORT_DATA = 'IMPORT_DATA';
export const UNDO_HISTORY = 'UNDO_HISTORY';
export const REDO_HISTORY = 'REDO_HISTORY';
export const RESET_HISTORY = 'RESET_HISTORY';
export const START_HISTORY_ITEM = 'START_HISTORY_ITEM';
export const FINISH_HISTORY_ITEM = 'FINISH_HISTORY_ITEM';
export const ADD_NEW_DRAGGABLE_POINT = 'ADD_NEW_DRAGGABLE_POINT';
export const REMOVE_DRAGGABLE_POINT = 'REMOVE_DRAGGABLE_POINT';
export const SET_SELECTED_ELEMENT = 'SET_SELECTED_ELEMENT';
export const SET_SELECTED_MODULE_SPEC = 'SET_SELECTED_MODULE_SPEC';
export const SET_TRANSFORMER_NODES = 'SET_TRANSFORMER_NODES';
export const CLEAR_TRANSFORMER_NODES_STATE = 'CLEAR_TRANSFORMER_NODES_STATE';
export const TOGGLE_MODULE = 'TOGGLE_MODULE';

export const defaultMultiplierScale = 10;

export interface SolarElementReference {
  layer?: string;
  id: string;
}

interface AddShapeAction {
  type: typeof ADD_SHAPE;
  payload: {
    element: SolarElement;
    moduleSeries: string;
  };
}

interface UpdateShapeAction {
  type: typeof UPDATE_SHAPE;
  payload: SolarElement;
}

interface RemoveShapeAction {
  type: typeof REMOVE_SHAPE;
  payload: {
    layer: string;
    elementIds: string[];
  };
}

interface ToggleVisibilityAction {
  type: typeof TOGGLE_VISIBILITY;
  payload: SolarElementReference;
}

interface ToggleLockAction {
  type: typeof TOGGLE_LOCK;
  payload: SolarElementReference;
}

interface ImportDataAction {
  type: typeof IMPORT_DATA;
  payload: SolarDesign;
}

interface UndoHistoryAction {
  type: typeof UNDO_HISTORY;
}

interface RedoHistoryAction {
  type: typeof REDO_HISTORY;
}

interface ResetHistoryAction {
  type: typeof RESET_HISTORY;
}

interface StartHistoryItemAction {
  type: typeof START_HISTORY_ITEM;
  payload: {
    ids: string[];
    layer: string;
  };
}

interface FinishHistoryItemAction {
  type: typeof FINISH_HISTORY_ITEM;
  payload: {
    ids: string[];
    layer: string;
  };
}

interface AddNewDraggablePoint {
  type: typeof ADD_NEW_DRAGGABLE_POINT;
  payload: {
    layer: string;
    elementId: string;
    newPoint: ShapeCoords;
  };
}

interface RemoveDraggablePoint {
  type: typeof REMOVE_DRAGGABLE_POINT;
  payload: {
    layer: string;
    elementId: string;
    pointIndex: number;
  };
}

interface SetSelectedElement {
  type: typeof SET_SELECTED_ELEMENT;
  payload: {
    layer: string;
    elementId: string;
    selected: boolean;
  };
}

interface SetSelectedModuleSpec {
  type: typeof SET_SELECTED_MODULE_SPEC;
  payload: ModuleSpec;
}

interface SetTransformerNodes {
  type: typeof SET_TRANSFORMER_NODES;
  payload: {
    selectionType: SelectionType;
    elementIds: string[];
  };
}

interface ClearTransformerNodesState {
  type: typeof CLEAR_TRANSFORMER_NODES_STATE;
}

interface ToggleModule {
  type: typeof TOGGLE_MODULE;
  payload: {
    elementId: string;
  };
}

export type StageActionTypes =
  | AddShapeAction
  | UpdateShapeAction
  | RemoveShapeAction
  | ToggleVisibilityAction
  | ToggleLockAction
  | ImportDataAction
  | UndoHistoryAction
  | RedoHistoryAction
  | ResetHistoryAction
  | StartHistoryItemAction
  | FinishHistoryItemAction
  | AddNewDraggablePoint
  | RemoveDraggablePoint
  | SetSelectedElement
  | SetSelectedModuleSpec
  | SetTransformerNodes
  | ClearTransformerNodesState
  | ToggleModule;
