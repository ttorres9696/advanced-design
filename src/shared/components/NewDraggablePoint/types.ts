import { Vector2d } from 'konva/types/types';

export interface NewDraggablePointProps {
  newDraggablePointMode: boolean;
  mousePosition?: Vector2d;
}
