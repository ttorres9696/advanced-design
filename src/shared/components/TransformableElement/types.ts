import Konva from 'konva';
import { Node } from 'konva/types/Node';
import { ReactElement } from 'react';

import { SelectionType } from '../../../redux/canvas/types';
import { ShapeElementType } from '../ShapeElement/types';

export interface UseTansfromableElementProps {
  id: string;
  layer: string;
  onStartChange: (id: string, layer: string) => void;
  onChange: (id: string, layer: string, params: TransformableElementChangeParams) => void;
  onFinishChange: (id: string, layer: string) => void;
  onRotate: (id: string, layer: string, pointsRotationInRadians: number[], points: number[][]) => void;
  setDragging: (dragging: boolean) => void;
  version: number;
  deleteDraggablePointMode: boolean;
  onMoveStart: (layer: string, node: Node) => void;
  onMove: (id: string, layer: string, node: Node, e: Konva.KonvaEventObject<DragEvent>) => void;
  onMoveEnd: (layer: string, node: Node) => void;
  onSelect: (id: string, layer: string, selectionType: SelectionType, node: Node) => void;
  visible: boolean;
  locked: boolean;
  multiple: boolean;
  selectMode: boolean;
}

export interface TransformableElementProps extends UseTansfromableElementProps {
  shapeType: ShapeElementType;
  resizeEnabled?: boolean;
  children?: ReactElement | ReactElement[];
  isSelected: boolean;
  active: boolean;
}

export interface TransformableElementChangeParams {
  x?: number;
  y?: number;
  scaleX?: number;
  scaleY?: number;
  width?: number;
  height?: number;
  radius?: number;
  points?: number[][];
  pointsRotationInRadians?: number[];
}

export interface TransformerDimensions {
  width: number;
  height: number;
}
