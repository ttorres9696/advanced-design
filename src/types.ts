import { CallbacksProps } from './shared/models/CallbacksProps.interface';
import { SolarDesign } from './shared/models/SolarDesign.interface';

export interface AdvancedDesignProps extends CallbacksProps {
  solarDesign: SolarDesign;
}
