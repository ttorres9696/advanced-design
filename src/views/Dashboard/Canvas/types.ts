import Konva from 'konva';
import { Node } from 'konva/types/Node';

import { SelectionType } from '../../../redux/canvas/types';
import { SolarDesign } from '../../../shared/models/SolarDesign.interface';

export interface CanvasProps {
  solarDesign: SolarDesign;
  canvasDimension: CanvasDimension;
}

export interface CanvasMargin {
  left?: number;
  top?: number;
  right?: number;
  bottom?: number;
}

export interface CanvasDimension {
  width: number;
  height: number;
}

export interface UseCanvasProps {
  solarDesign: SolarDesign;
  transformerSelectElement: (selectionType: SelectionType, node?: Node) => void;
  transformerHandleDragElement: (node: Node, e: Konva.KonvaEventObject<DragEvent>) => void;
  transformerOnDragEnd: (layer: string, node: Node) => string[];
}
