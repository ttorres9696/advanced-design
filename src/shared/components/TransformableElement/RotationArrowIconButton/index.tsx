import { Image as ImageType } from 'konva/types/shapes/Image';
import { Stage } from 'konva/types/Stage';
import React, { Fragment, useCallback, useEffect, useRef, useState } from 'react';
import { Image } from 'react-konva';

import { redoBlack } from '../../../assets/redoBlack';
import { undoBlack } from '../../../assets/undoBlack';
import { RotationArrowIconButtonProps, RotationIconSize } from './types';

const RotationArrowIconButton: React.FC<RotationArrowIconButtonProps> = ({
  x,
  y,
  position,
  onClick,
  type,
  rotation,
  invisible,
}) => {
  const shapeRef = useRef<any>();

  const [mouseOver, setMouseOver] = useState<boolean>(false);
  const [icon, setIcon] = useState<HTMLImageElement>();
  const [loadedIcon, setLoadedIcon] = useState<boolean>(false);

  const onClickIcon = useCallback((e: any) => onClick(position), [onClick, position]);

  const onMouseOver = useCallback(
    (e: any) => {
      setMouseOver(true);
    },
    [setMouseOver],
  );

  const onMouseOut = useCallback((e: any) => setMouseOver(false), [setMouseOver]);

  const onMouseEnter = useCallback(() => {
    if (shapeRef && shapeRef.current) {
      const node = shapeRef.current as ImageType;
      const stage: Stage | null = node.getStage();

      if (stage) {
        stage.container().style.cursor = 'pointer';
      }
    }
  }, [shapeRef]);

  const onMouseLeave = useCallback(() => {
    if (shapeRef && shapeRef.current) {
      const node = shapeRef.current as ImageType;
      const stage: Stage | null = node.getStage();

      if (stage) {
        stage.container().style.cursor = 'default';
      }
    }
  }, [shapeRef]);

  const handleImageLoad = useCallback(() => {
    setLoadedIcon(true);
  }, [setLoadedIcon]);

  useEffect(() => {
    if (!icon) {
      const image = new window.Image();
      image.src = type === 'left' ? undoBlack : redoBlack;
      image.addEventListener('load', handleImageLoad);

      setIcon(image);
    }
  }, [icon, setIcon, redoBlack, undoBlack, handleImageLoad]);

  return (
    <Fragment>
      {loadedIcon ? (
        <Image
          ref={shapeRef}
          image={icon}
          x={x}
          y={y}
          width={RotationIconSize}
          height={RotationIconSize}
          tension={0}
          opacity={mouseOver ? 1 : 0.5}
          draggable={false}
          rotation={rotation}
          onClick={onClickIcon}
          onMouseOver={onMouseOver}
          onMouseOut={onMouseOut}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
          visible={!invisible}
        />
      ) : null}
    </Fragment>
  );
};

export default React.memo(RotationArrowIconButton);
