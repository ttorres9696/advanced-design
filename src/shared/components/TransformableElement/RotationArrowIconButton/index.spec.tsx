import { mount, ReactWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Layer, Stage } from 'react-konva';

import RotationArrowIconButton from '.';
import { RotationArrowIconButtonProps } from './types';

let wrapper: ReactWrapper;

beforeEach(() => {
  const props: RotationArrowIconButtonProps = {
    x: 100,
    y: 100,
    position: 0,
    onClick: (position: number) => {},
    type: 'left',
    rotation: 0,
  };

  wrapper = mount(
    <Stage>
      <Layer>
        <RotationArrowIconButton {...props} />
      </Layer>
    </Stage>,
    {},
  );
});

describe('RotationArrowIconButton', () => {
  it('renders correctly', () => expect(toJson(wrapper)).toMatchSnapshot());
});
