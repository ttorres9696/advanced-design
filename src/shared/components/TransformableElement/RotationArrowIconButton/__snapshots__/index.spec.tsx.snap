// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RotationArrowIconButton renders correctly 1`] = `
<ForwardRef>
  <Stage
    forwardedRef={null}
  >
    <div />
  </Stage>
</ForwardRef>
`;
